posts/python-decorators.md

```diff
@@ -0,0 +1,180 @@
+        f()
+    return log_f_as_called
+
+@mydecorator
+@mydecorator
+def hello():
+    print('hello')
+
+>>> hello()
+called <function hello at 0x7f27737c7950>
+called <function hello at 0x7f27737c7f28>
+hello
+```
+
+Now, our new function appears just like the one its wrapping/decorating. However, we are still reliant on the fact that it returns nothing, and accepts no input. If we wanted to make that more general, we would need to pass in the arguments and return the same value. We can modify our function to look like this:
+
+
+```python
+import functools
+def mydecorator(f): 
+    @functools.wraps(f)  # we tell wraps that the function we are wrapping is f
+    def log_f_as_called(*args, **kwargs):
+        print(f'{f} was called with arguments={args} and kwargs={kwargs}')
+        value = f(*args, **kwargs)
+        print(f'{f} return value {value}')
+        return value
+    return log_f_as_called
+```
+
+Now we are printing everytime the function is called, including all the inputs the function receives, and what it returns. Now you can simply decorate any existing function and get free debug logging, without needing to change the function at all. 
+
+# Adding variables to the decorator
+If you were using the decorator for any code that you wanted to ship, and not just for yourself locally, you might want to replace all the `print` statements with logging statements. In which case, you would need to define a log level. It might be safe to assume you always want the debug log level, but what that also might depend on the function. We can provide variables to the decorator itself, to define how it should behave. For example:
+
+```python
+@debug(level='info')
+def hello():
+    print('hello')
+```
+The above code would allow us to specify that this specific function should log at the info level instead of the debug level. This is achieved in python by writing a function, that returns a decorator. Yes, a decorator is also a function. So this is essentially saying `hello = debug('info')(hello)`. That double parenthesis might look funky, but basically, debug is function, that returns a function. Adding this to our existing decorator, we would need to nest one more time, now making our code look like the following:
+
+```python
+import functools
+def debug(level): 
+    def mydecorator(f)
+        @functools.wraps(f)   # we tell wraps that the function we are wrapping is f
+        def log_f_as_called(*args, **kwargs):
+            logger.log(level, f'{f} was called with arguments={args} and kwargs={kwargs}')
+            value = f(*args, **kwargs)
+            logger.log(level, f'{f} return value {value}')
+            return value
+        return log_f_as_called
+    return mydecorator
+```
+
+This is starting to get a bit ugly, and overly nested. A cool little trick to solve that, which I like to do, is simply add a default kwarg `level` to `debug` and return a [partial](https://docs.python.org/2/library/functools.html#functools.partial).
```

Comment:
I would try to explain this a little more in-depth. What is a partial? Why did you make the changes you did?